examples/recursive-types-2/F.hs

```diff
@@ -0,0 +1,100 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- | This is similar to the @recursive-types@ example, except that the @Tree@ type
+-- is monomorphic. This makes it slightly more complicated, because in the other
+-- example, we could define the @TargetOb@ instance for @Tree@ as
+--
+--  @
+--     type instance TargetOb (Tree a) = Tree (TargetOb a)
+--  @
+--
+-- while here we have to create another data type, @TreeAux@, and use it to define
+-- @TargetOb Tree@.
+--
+-- Note that the following won't work:
+--
+--  @
+--     -- Won't work because Tree does not have CGeneric instance
+--     type instance TargetOb Tree = TargetOb (CG.Rep Tree ())
+--
+--     -- Won't work because this doesn't terminate (Tree is recursive)
+--     type instance TargetOb Tree = TargetOb (Categorifier.Client.Rep Tree)
+--  @
+--
+-- The generated C code is the same as the @recursive-types@ eaxmple.
+module F (fCategorified) where
+
+import qualified Categorifier.C.CExpr.Cat as C
+import Categorifier.C.CExpr.Cat.TargetOb (TargetOb)
+import Categorifier.C.CTypes.CGeneric (CGeneric)
+import qualified Categorifier.C.CTypes.CGeneric as CG
+import Categorifier.C.CTypes.GArrays (GArrays)
+import Categorifier.C.KTypes.C (C)
+import qualified Categorifier.Categorify as Categorify
+import qualified Categorifier.Client as Client
+import Data.Int (Int32, Int64)
+import GHC.Generics (Generic)
+
+data Input = Input
+  { iInt32 :: C Int32,
+    iInt64 :: C Int64
+  }
+  deriving (Generic)
+
+Client.deriveHasRep ''Input
+
+instance CGeneric Input
+
+instance GArrays C Input
+
+type instance TargetOb Input = TargetOb (CG.Rep Input ())
+
+data Output = Output
+  { oInt64 :: C Int64,
+    oBool :: Bool
+  }
+  deriving (Generic)
+
+Client.deriveHasRep ''Output
+
+instance CGeneric Output
+
+instance GArrays C Output
+
+type instance TargetOb Output = TargetOb (CG.Rep Output ())
+
+data Tree = Empty | Branch (C Int32) Tree Tree
+  deriving (Generic)
+
+Client.deriveHasRep ''Tree
+
+data TreeAux = EmptyAux | BranchAux (TargetOb Int32) TreeAux TreeAux
+  deriving (Generic)
+
+Client.deriveHasRep ''TreeAux
+
+type instance TargetOb Tree = TreeAux
```

Comment:
I thought for this sort of type the `TargetOb` was like `TargetOb (Client.Rep Tree)`? Or was it `CGRep`. I'm guessing that doesn't work here because it's not “deep”? You should add a comment explaining why the usual forms don't work in this case.

Reply:
It's explained in the haddock above. The `Rep` one doesn't terminate because `Tree` is recursive. And there's no `CGeneric` instance for `Tree` (at least the default instance doesn't work).

Reply:
Whoops! Perfect.